File: lib/call-state.ts
import type { CallStatus } from "@/types";

export const CALL_TIMEOUT_MS = 40_000;

const allowedTransitions: Record<CallStatus, CallStatus[]> = {
  idle: ["incoming_ringing", "acquiring_media"],
  incoming_ringing: ["acquiring_media", "ending", "failed"],
  outgoing_ringing: ["connecting", "ending", "failed"],
  acquiring_media: ["outgoing_ringing", "connecting", "ending", "failed"],
  connecting: ["connected", "reconnecting", "ending", "failed"],
  connected: ["reconnecting", "ending", "failed"],
  reconnecting: ["connected", "ending", "failed"],
  ending: ["ended", "failed"],
  ended: ["idle"],
  failed: ["idle", "ending"]
};

export function canTransitionCall(from: CallStatus, to: CallStatus) {
  if (from === to) return true;
  return allowedTransitions[from]?.includes(to) ?? false;
}

export function readableCallStatus(status: CallStatus) {
  return status.replace(/_/g, " ");
}

File: lib/features/calls/domain/call_state.dart
enum CommsCallStatus {
  idle,
  outgoingRinging,
  incomingRinging,
  acquiringMedia,
  connecting,
  connected,
  reconnecting,
  ending,
  ended,
  failed,
}

const callTimeout = Duration(seconds: 45);

const allowedCallTransitions = <CommsCallStatus, Set<CommsCallStatus>>{
  CommsCallStatus.idle: {
    CommsCallStatus.incomingRinging,
    CommsCallStatus.outgoingRinging,
    CommsCallStatus.acquiringMedia
  },
  CommsCallStatus.incomingRinging: {
    CommsCallStatus.acquiringMedia,
    CommsCallStatus.ending,
    CommsCallStatus.ended,
    CommsCallStatus.failed
  },
  CommsCallStatus.outgoingRinging: {
    CommsCallStatus.acquiringMedia,
    CommsCallStatus.connecting,
    CommsCallStatus.ending,
    CommsCallStatus.ended,
    CommsCallStatus.failed
  },
  CommsCallStatus.acquiringMedia: {
    CommsCallStatus.outgoingRinging,
    CommsCallStatus.connecting,
    CommsCallStatus.ending,
    CommsCallStatus.failed
  },
  CommsCallStatus.connecting: {
    CommsCallStatus.connected,
    CommsCallStatus.reconnecting,
    CommsCallStatus.ending,
    CommsCallStatus.ended,
    CommsCallStatus.failed
  },
  CommsCallStatus.connected: {
    CommsCallStatus.reconnecting,
    CommsCallStatus.ending,
    CommsCallStatus.ended,
    CommsCallStatus.failed
  },
  CommsCallStatus.reconnecting: {
    CommsCallStatus.acquiringMedia,
    CommsCallStatus.connecting,
    CommsCallStatus.connected,
    CommsCallStatus.ending,
    CommsCallStatus.ended,
    CommsCallStatus.failed
  },
  CommsCallStatus.ending: {CommsCallStatus.ended, CommsCallStatus.failed},
  CommsCallStatus.ended: {CommsCallStatus.idle},
  CommsCallStatus.failed: {CommsCallStatus.idle, CommsCallStatus.ending},
};

bool canTransitionCall(CommsCallStatus from, CommsCallStatus to) {
  return from == to || (allowedCallTransitions[from]?.contains(to) ?? false);
}
